afl-fuzz.c

```diff
@@ -1169,61 +1032,88 @@ EXP_ST void init_count_class16(void) {
 
 }
 
+/* Import coverage processing routines. */
 
 #ifdef WORD_SIZE_64
+#  include "coverage-64.h"
+#else
+#  include "coverage-32.h"
+#endif
 
-static inline void classify_counts(u64* mem) {
 
-  u32 i = MAP_SIZE >> 3;
+/* Check if the current execution path brings anything new to the table.
+   Update virgin bits to reflect the finds. Returns 1 if the only change is
+   the hit-count for a particular tuple; 2 if there are new tuples seen. 
+   Updates the map, so subsequent calls will always return 0.
 
-  while (i--) {
+   This function is called after every exec() on a fairly large buffer, so
+   it needs to be fast. We do this in 32-bit and 64-bit flavors. */
 
-    /* Optimize for sparse bitmaps. */
+static inline u8 has_new_bits(u8* virgin_map) {
 
-    if (unlikely(*mem)) {
+#ifdef WORD_SIZE_64
 
-      u16* mem16 = (u16*)mem;
+  u64* current = (u64*)trace_bits;
+  u64* virgin  = (u64*)virgin_map;
 
-      mem16[0] = count_class_lookup16[mem16[0]];
-      mem16[1] = count_class_lookup16[mem16[1]];
-      mem16[2] = count_class_lookup16[mem16[2]];
-      mem16[3] = count_class_lookup16[mem16[3]];
+  u32  i = (MAP_SIZE >> 3);
 
-    }
+#else
 
-    mem++;
+  u32* current = (u32*)trace_bits;
+  u32* virgin  = (u32*)virgin_map;
 
-  }
+  u32  i = (MAP_SIZE >> 2);
 
-}
+#endif /* ^WORD_SIZE_64 */
 
-#else
+  u8   ret = 0;
 
-static inline void classify_counts(u32* mem) {
+  while (i--) {
 
-  u32 i = MAP_SIZE >> 2;
+    if (unlikely(*current)) discover_word(&ret, current, virgin);
 
-  while (i--) {
+    current++;
+    virgin++;
 
-    /* Optimize for sparse bitmaps. */
+  }
 
-    if (unlikely(*mem)) {
+  if (ret && virgin_map == virgin_bits) bitmap_changed = 1;
 
-      u16* mem16 = (u16*)mem;
+  return ret;
 
-      mem16[0] = count_class_lookup16[mem16[0]];
-      mem16[1] = count_class_lookup16[mem16[1]];
+}
 
-    }
 
-    mem++;
+/* A combination of classify_counts and has_new_bits. If 0 is returned, then the
+ * trace bits are kept as-is. Otherwise, the trace bits are overwritten with
+ * classified values.
+ *
+ * This accelerates the processing: in most cases, no interesting behavior
+ * happen, and the trace bits will be discarded soon. This function optimizes
+ * for such cases: one-pass scan on trace bits without modifying anything. Only
+ * on rare cases it fall backs to the slow path: classify_counts() first, then
+ * return has_new_bits(). */
```

Comment:
Maybe mention all return values. 0 is actually overloaded, representing "no new bits" and "trace bits not classified".

Instead of referencing the previous version (combining, accelerating, etc), make the comment self-contained.